File: Matrox Imaging Library (MIL) X 23H2 (10.60)/Examples/Full/MgraInteractive/C++/MgraInteractive.cpp
/*****************************************************************************/
/* 
 * File name: MGraInteractive.cpp
 *
 * Synopsis:  This program uses the capabilities of MIL's interactive
 *            graphics, along with the Blob analysis module, to count 
 *            objects within a user-defined region.
 *
 * Copyright © Matrox Electronic Systems Ltd., 1992-2023.
 * All Rights Reserved
 */

#include <mil.h>

/* Data structure to handle parameters for the hook function. */
struct STestParameters
   {
   MIL_ID   MilDisplay,
            MilGraphicsList,
            MilGraphicsContext,
            MilBinImage,
            MilBlobContext,
            MilBlobResult;
   MIL_INT  RegionLabel;
   };

/* Example function declarations. */ 
void           InteractivityExample(MIL_ID MilSystem, MIL_ID MilDisplay);

MIL_INT MFTYPE HookHandler(MIL_INT HookType, MIL_ID EventId, void* UserData);

void           CountObjects(MIL_ID MilDisplay, MIL_ID MilGraphicsList,
                            MIL_ID MilGraphicsContext, MIL_ID MilBinImage,
                            MIL_ID MilBlobContext, MIL_ID MilBlobResult);

/*****************************************************************************
 Main.
 *****************************************************************************/
int MosMain(void)
   {
   MIL_ID MilApplication,     /* Application identifier. */
          MilSystem,          /* System Identifier.      */
          MilDisplay;         /* Display identifier.     */

   /* Allocate defaults. */
   MappAllocDefault(M_DEFAULT, &MilApplication, &MilSystem, &MilDisplay, 
      M_NULL, M_NULL);

   /* Print example name. */ 
   MosPrintf(MIL_TEXT("\nINTERACTIVE REGIONS AND SUBPIXEL ANNOTATIONS:\n"));
   MosPrintf(MIL_TEXT("---------------------------------------------\n\n"));
   MosPrintf(MIL_TEXT("This program determines the number of blobs in a region\n"));
   MosPrintf(MIL_TEXT("defined interactively by the user. The extracted blob's\n"));
   MosPrintf(MIL_TEXT("features are drawn with subpixel accuracy in a zoomable\n"));
   MosPrintf(MIL_TEXT("display.\n\n"));

   /* Run Interactivity Example. */
   InteractivityExample(MilSystem, MilDisplay);
 
   /* Free defaults. */    
   MappFreeDefault(MilApplication, MilSystem, MilDisplay, M_NULL, M_NULL);
   
   return 0;
   }

/*****************************************************************************/
/* Interactivity example. */

/* Source MIL image file specification. */ 
#define IMAGE_FILE M_IMAGE_PATH MIL_TEXT("Seals.mim")
/* Threshold for image binarization. */
#define IMAGE_THRESHOLD_VALUE 110L 
/* Initial region parameters. */
#define RECTANGLE_POSITION_X 160
#define RECTANGLE_POSITION_Y 310
#define RECTANGLE_WIDTH      200
#define RECTANGLE_HEIGHT     175
#define RECTANGLE_ANGLE      0
/* Interactivity parameters. */
#define SELECTION_RADIUS     10
/* Value of enter key in Ascii table */
#define ASCII_ENTER 13

void InteractivityExample(MIL_ID MilSystem, MIL_ID MilDisplay)
   {
   MIL_ID  MilImage,                  /* Image buffer identifier.        */
           MilGraphicsList,           /* Graphics list identifier.       */
           MilGraphicsContext,        /* Graphics context identifier.    */
           MilBinImage,               /* Binary image buffer identifier. */
           MilBlobContext,        	  /* Context identifier.             */
           MilBlobResult;             /* Blob result buffer identifier.  */

   MIL_INT SizeX,                     /* Size X of the source buffer.    */
           SizeY,                     /* Size Y of the source buffer.    */
           RegionLabel;               /* Label value of the region.      */

   STestParameters DataStructure;     /* Hook function data structure.   */
   
   /* Restore the source image. */
   MbufRestore(IMAGE_FILE, MilSystem, &MilImage);

   /* Display the buffer. */
   MdispSelect(MilDisplay, MilImage);

   /* Allocate a graphics list to hold the subpixel annotations. */
   MgraAllocList(MilSystem, M_DEFAULT, &MilGraphicsList);

   /* Increase the selection radius for easier interactivity. */
   MgraControlList(MilGraphicsList, M_LIST, M_DEFAULT, M_SELECTION_RADIUS,
      SELECTION_RADIUS);

   /* Associate the graphics list to the display for annotations. */
   MdispControl(MilDisplay, M_ASSOCIATED_GRAPHIC_LIST_ID, MilGraphicsList);

   /* Allocate a graphics context for the draw operations. */
   MgraAlloc(MilSystem, &MilGraphicsContext);

   /* Enable the interactive mode. */
   MdispControl(MilDisplay, M_GRAPHIC_LIST_INTERACTIVE, M_ENABLE);

   /* Enable the use of action keys */
   MgraControlList(MilGraphicsList, M_LIST, M_DEFAULT, M_ACTION_KEYS, M_ENABLE);

   /* Add a selectable rectangular region.*/
   MgraRectAngle(MilGraphicsContext, MilGraphicsList, RECTANGLE_POSITION_X, 
      RECTANGLE_POSITION_Y, RECTANGLE_WIDTH, RECTANGLE_HEIGHT, RECTANGLE_ANGLE,
      M_CENTER_AND_DIMENSION);
   
   /* Retrieve the label of the rectangle graphic. */
   MgraInquireList(MilGraphicsList, M_LIST, M_DEFAULT, M_LAST_LABEL, &RegionLabel);

   /* Modify the line thickness of the rectangle. */
   MgraControlList(MilGraphicsList, M_GRAPHIC_LABEL(RegionLabel), M_DEFAULT, M_LINE_THICKNESS, 5.0);

   /* Disable the selectable mode for the next annotations to the graphics list. */
   MgraControl(MilGraphicsContext, M_SELECTABLE, M_DISABLE);

   /* Allocate a binary image buffer for fast processing. */
   MbufInquire(MilImage, M_SIZE_X, &SizeX);
   MbufInquire(MilImage, M_SIZE_Y, &SizeY);
   MbufAlloc2d(MilSystem, SizeX, SizeY, 1+M_UNSIGNED, M_IMAGE+M_PROC, &MilBinImage);
 
   /* Binarize the source image. */
   MimBinarize(MilImage, MilBinImage, M_FIXED+M_LESS, IMAGE_THRESHOLD_VALUE, M_NULL);
   
   /* Allocate a blob context and a blob result. */ 
   MblobAlloc(MilSystem, M_DEFAULT, M_DEFAULT, &MilBlobContext);
   MblobAllocResult(MilSystem, M_DEFAULT, M_DEFAULT, &MilBlobResult);

   /* Select the blob features to calculate (Center Of Gravity and Box). */
   MblobControl(MilBlobContext, M_CENTER_OF_GRAVITY + M_BINARY, M_ENABLE);
   MblobControl(MilBlobContext, M_BOX, M_ENABLE);

   /* Programmatically initialize the selected state of the rectangle region. */
   MgraControlList(MilGraphicsList, M_GRAPHIC_LABEL(RegionLabel), M_DEFAULT, 
      M_GRAPHIC_SELECTED, M_TRUE);

   /* Perform and display a first count of the number of objects */
   /* within the initial region.                                 */
   CountObjects(MilDisplay, MilGraphicsList, MilGraphicsContext, MilBinImage,
      MilBlobContext, MilBlobResult);

   /* Initialize the hook data structure, then associate the hook function to */ 
   /* the "M_GRAPHIC_MODIFIED" event. The hook function will be called        */
   /* with any region interaction by the user.                                */
   DataStructure.MilDisplay         = MilDisplay;
   DataStructure.MilGraphicsList    = MilGraphicsList;
   DataStructure.MilGraphicsContext = MilGraphicsContext;
   DataStructure.MilBinImage        = MilBinImage;
   DataStructure.RegionLabel        = RegionLabel;
   DataStructure.MilBlobContext     = MilBlobContext;
   DataStructure.MilBlobResult      = MilBlobResult;
      
   MgraHookFunction(MilGraphicsList, M_GRAPHIC_MODIFIED, &HookHandler, &DataStructure);

   MosPrintf(MIL_TEXT("You can try using your mouse or your keyboard to interactively\n"));
   MosPrintf(MIL_TEXT("modify the displayed region, such as moving, resizing, or\n"));
   MosPrintf(MIL_TEXT("rotating the region. If you do so, the results and annotations\n"));
   MosPrintf(MIL_TEXT("will be immediately updated.\n\n"));

   MosPrintf(MIL_TEXT("Press <Enter> to exit.\n"));
   MIL_INT PressedKey = 0;
   while(PressedKey != ASCII_ENTER)
      PressedKey = MosGetch();

   MgraHookFunction(MilGraphicsList, M_GRAPHIC_MODIFIED+M_UNHOOK,
      &HookHandler, &DataStructure);

   /* Free all allocations. */
   MblobFree(MilBlobResult); 
   MblobFree(MilBlobContext); 
   MbufFree(MilBinImage);
   MgraFree(MilGraphicsContext);
   MgraFree(MilGraphicsList);
   MbufFree(MilImage);
   }


MIL_INT MFTYPE HookHandler(MIL_INT HookType, MIL_ID EventId, void* UserData)
   {
   STestParameters* pDataStructure = (STestParameters*)(UserData);

   /* Check that the modified graphic is the rectangular region. */
   MIL_INT ModifiedGraphicLabel;
   MgraGetHookInfo(EventId, M_GRAPHIC_LABEL_VALUE, &ModifiedGraphicLabel);

   if (ModifiedGraphicLabel == pDataStructure->RegionLabel)
      {
      /* Count objects and draw the corresponding annotations. */
      CountObjects(pDataStructure->MilDisplay,
                   pDataStructure->MilGraphicsList, 
                   pDataStructure->MilGraphicsContext,
                   pDataStructure->MilBinImage, 
                   pDataStructure->MilBlobContext, 
                   pDataStructure->MilBlobResult);
      }

   return M_NULL;
   }


#define MAX_TEXT_SIZE 100

void CountObjects(MIL_ID MilDisplay, MIL_ID MilGraphicsList, MIL_ID MilGraphicsContext,
                  MIL_ID MilBinImage, MIL_ID MilBlobContext, MIL_ID MilBlobResult)
   {
   MIL_INT NumberOfBlobs,
           NumberOfPrimitives,
           Index;

   MIL_TEXT_CHAR TextLabel[MAX_TEXT_SIZE];

   /* Disable the display update for better performance. */
   MdispControl(MilDisplay, M_UPDATE, M_DISABLE);
      
   /* Remove all elements from the graphics list, except the rectangle */
   /* region primitive at index 0.                                     */ 
   MgraInquireList(MilGraphicsList, M_LIST, M_DEFAULT, M_NUMBER_OF_GRAPHICS,
      &NumberOfPrimitives);
   for(Index = NumberOfPrimitives-1; Index > 0; Index--)
      {
      MgraControlList(MilGraphicsList, M_GRAPHIC_INDEX(Index), M_DEFAULT, M_DELETE,
         M_DEFAULT);
      }

   /* Set the input region. The blob analysis will be done */
   /* from the (filled) interactive rectangle.             */
   MbufSetRegion(MilBinImage, MilGraphicsList, M_DEFAULT,
      M_RASTERIZE + M_FILL_REGION + M_USE_LINE_THICKNESS_1, M_DEFAULT);
      
   /* Calculate the blobs and their features. */ 
   MblobCalculate(MilBlobContext, MilBinImage, M_NULL, MilBlobResult);

   /* Get the total number of blobs. */ 
   MblobGetResult(MilBlobResult, M_GENERAL, M_NUMBER + M_TYPE_MIL_INT, &NumberOfBlobs);

   /* Set the input units to display unit for the count annotations. */ 
   MgraControl(MilGraphicsContext, M_INPUT_UNITS, M_DISPLAY);
   MosSprintf(TextLabel, MAX_TEXT_SIZE, MIL_TEXT(" Number of blobs found: %2i "),
      (int) NumberOfBlobs);

   MgraColor(MilGraphicsContext, M_COLOR_WHITE);
   MgraText(MilGraphicsContext, MilGraphicsList, 10, 10, TextLabel);
      
   /* Restore the input units to pixel units for result annotations. */ 
   MgraControl(MilGraphicsContext, M_INPUT_UNITS, M_PIXEL);

   /* Draw blob center of gravity annotation. */
   MgraControl(MilGraphicsContext, M_LINE_THICKNESS, 3.0);
   MgraColor(MilGraphicsContext, M_COLOR_RED);
   MblobDraw(MilGraphicsContext, MilBlobResult, MilGraphicsList, M_DRAW_CENTER_OF_GRAVITY, 
      M_INCLUDED_BLOBS, M_DEFAULT);

   /* Draw blob bounding box annotations. */
   MgraColor(MilGraphicsContext, M_COLOR_GREEN);
   MgraControl(MilGraphicsContext, M_LINE_THICKNESS, 1.0);
   MblobDraw(MilGraphicsContext, MilBlobResult, MilGraphicsList, M_DRAW_BOX, 
      M_INCLUDED_BLOBS, M_DEFAULT);

   /* Enable the display to update the drawings. */
   MdispControl(MilDisplay, M_UPDATE, M_ENABLE);
   }

File: Matrox Imaging Library (MIL) X 23H2 (10.60)/Examples/Full/MgraInteractive/C#/MgraInteractive.cs
//*****************************************************************************
// 
// File name: MgraInteractive.cs
//
// Synopsis:  This program uses the capabilities of MIL's interactive
//            graphics, along with the Blob analysis module, to count 
//            objects within a user-defined region.
//
// Copyright © Matrox Electronic Systems Ltd., 1992-2023.
// All Rights Reserved
//*****************************************************************************
using System;
using System.Collections.Generic;
using System.Text;

using Matrox.MatroxImagingLibrary;
using System.Runtime.InteropServices;

namespace MGraInteractive
{
    public class STestParameters
    {
        public MIL_ID MilDisplay;
        public MIL_ID MilGraphicsList;
        public MIL_ID MilGraphicsContext;
        public MIL_ID MilBinImage;
        public MIL_ID MilBlobContext;
        public MIL_ID MilBlobResult;
        public MIL_INT RegionLabel;
    }

    public class Program
    {
        static void Main(string[] args)
        {
            MIL_ID MilApplication = MIL.M_NULL;     // Application identifier.
            MIL_ID MilSystem = MIL.M_NULL;          // System Identifier.
            MIL_ID MilDisplay = MIL.M_NULL;         // Display identifier.

            // Allocate defaults.
            MIL.MappAllocDefault(MIL.M_DEFAULT, ref MilApplication, ref MilSystem, ref MilDisplay, MIL.M_NULL, MIL.M_NULL);

            // Print example name.
            Console.WriteLine();
            Console.WriteLine("INTERACTIVE REGIONS AND SUBPIXEL ANNOTATIONS:");
            Console.WriteLine("---------------------------------------------");
            Console.WriteLine();
            Console.WriteLine("This program determines the number of blobs in a region");
            Console.WriteLine("defined interactively by the user. The extracted blob's");
            Console.WriteLine("features are drawn with subpixel accuracy in a zoomable");
            Console.WriteLine("display.");
            Console.WriteLine();

            // Run Interactivity Example.
            InteractivityExample(MilSystem, MilDisplay);

            // Free defaults.
            MIL.MappFreeDefault(MilApplication, MilSystem, MilDisplay, MIL.M_NULL, MIL.M_NULL);
        }

        //*****************************************************************************
        // Interactivity example.

        // Source MIL image file specification.
        private static readonly string IMAGE_FILE = MIL.M_IMAGE_PATH + "Seals.mim";

        // Threshold for image binarization.
        private const int IMAGE_THRESHOLD_VALUE = 110;

        // Initial region parameters.
        private const int RECTANGLE_POSITION_X = 160;
        private const int RECTANGLE_POSITION_Y = 310;
        private const int RECTANGLE_WIDTH = 200;
        private const int RECTANGLE_HEIGHT = 175;
        private const int RECTANGLE_ANGLE = 0;
        // Value of enter key in Ascii table
        private const int ASCII_ENTER = 13; 

        private static void InteractivityExample(MIL_ID MilSystem, MIL_ID MilDisplay)
        {
            MIL_ID MilImage = MIL.M_NULL;                           // Image buffer identifier.
            MIL_ID MilGraphicsList = MIL.M_NULL;                    // Graphics list identifier.
            MIL_ID MilGraphicsContext = MIL.M_NULL;                 // Graphics context identifier.
            MIL_ID MilBinImage = MIL.M_NULL;                        // Binary image buffer identifier.
            MIL_ID MilBlobContext = MIL.M_NULL;                     // Context identifier.
            MIL_ID MilBlobResult = MIL.M_NULL;                      // Blob result buffer identifier.

            MIL_INT SizeX = 0;                                      // Size X of the source buffer.
            MIL_INT SizeY = 0;                                      // Size Y of the source buffer.
            MIL_INT RegionLabel = 0;                                // Label value of the region.

            STestParameters DataStructure = new STestParameters();  // Hook function data structure.

            // Restore the source image.
            MIL.MbufRestore(IMAGE_FILE, MilSystem, ref MilImage);

            // Display the buffer.
            MIL.MdispSelect(MilDisplay, MilImage);

            // Allocate a graphics list to hold the subpixel annotations.
            MIL.MgraAllocList(MilSystem, MIL.M_DEFAULT, ref MilGraphicsList);

            // Associate the graphics list to the display for annotations.
            MIL.MdispControl(MilDisplay, MIL.M_ASSOCIATED_GRAPHIC_LIST_ID, MilGraphicsList);

            // Allocate a graphics context for the draw operations.
            MIL.MgraAlloc(MilSystem, ref MilGraphicsContext);

            // Enable the interactive mode.
            MIL.MdispControl(MilDisplay, MIL.M_GRAPHIC_LIST_INTERACTIVE, MIL.M_ENABLE);

            // Enable the use of action keys.
            MIL.MgraControlList(MilGraphicsList, MIL.M_LIST, MIL.M_DEFAULT, MIL.M_ACTION_KEYS, MIL.M_ENABLE);

            // Add a selectable rectangular region.
            MIL.MgraRectAngle(MilGraphicsContext, MilGraphicsList, RECTANGLE_POSITION_X, RECTANGLE_POSITION_Y, RECTANGLE_WIDTH, RECTANGLE_HEIGHT, RECTANGLE_ANGLE, MIL.M_CENTER_AND_DIMENSION);

            // Retrieve the label of the rectangle graphic.
            MIL.MgraInquireList(MilGraphicsList, MIL.M_LIST, MIL.M_DEFAULT, MIL.M_LAST_LABEL, ref RegionLabel);

            // Modify the line thickness of the rectangle.
            MIL.MgraControlList(MilGraphicsList, MIL.M_GRAPHIC_LABEL(RegionLabel),MIL.M_DEFAULT,MIL.M_LINE_THICKNESS, 5.0);

            // Disable the selectable mode for the next annotations to the graphics list.
            MIL.MgraControl(MilGraphicsContext, MIL.M_SELECTABLE, MIL.M_DISABLE);

            // Allocate a binary image buffer for fast processing.
            MIL.MbufInquire(MilImage, MIL.M_SIZE_X, ref SizeX);
            MIL.MbufInquire(MilImage, MIL.M_SIZE_Y, ref SizeY);
            MIL.MbufAlloc2d(MilSystem, SizeX, SizeY, 1 + MIL.M_UNSIGNED, MIL.M_IMAGE + MIL.M_PROC, ref MilBinImage);

            // Binarize the source image.
            MIL.MimBinarize(MilImage, MilBinImage, MIL.M_FIXED + MIL.M_LESS, IMAGE_THRESHOLD_VALUE, MIL.M_NULL);

            // Allocate a blob context and a blob result.
            MIL.MblobAlloc(MilSystem, MIL.M_DEFAULT, MIL.M_DEFAULT, ref MilBlobContext);
            MIL.MblobAllocResult(MilSystem, MIL.M_DEFAULT, MIL.M_DEFAULT, ref MilBlobResult);

            // Select the blob features to calculate (Center Of Gravity and Box).
            MIL.MblobControl(MilBlobContext, MIL.M_CENTER_OF_GRAVITY + MIL.M_BINARY, MIL.M_ENABLE);
            MIL.MblobControl(MilBlobContext, MIL.M_BOX, MIL.M_ENABLE);

            // Programmatically initialize the selected state of the rectangle region.
            MIL.MgraControlList(MilGraphicsList, MIL.M_GRAPHIC_LABEL(RegionLabel), MIL.M_DEFAULT, MIL.M_GRAPHIC_SELECTED, MIL.M_TRUE);

            // Perform and display a first count of the number of objects
            // within the initial region.                                
            CountObjects(MilDisplay, MilGraphicsList, MilGraphicsContext, MilBinImage, MilBlobContext, MilBlobResult);

            // Initialize the hook data structure, then associate the hook function to
            // the "MIL.M_GRAPHIC_MODIFIED" event. The hook function will be called
            // with any region interaction by the user.
            DataStructure.MilDisplay = MilDisplay;
            DataStructure.MilGraphicsList = MilGraphicsList;
            DataStructure.MilGraphicsContext = MilGraphicsContext;
            DataStructure.MilBinImage = MilBinImage;
            DataStructure.RegionLabel = RegionLabel;
            DataStructure.MilBlobContext = MilBlobContext;
            DataStructure.MilBlobResult = MilBlobResult;

            GCHandle DataStructureHandle = GCHandle.Alloc(DataStructure);
            MIL_GRA_HOOK_FUNCTION_PTR HookHandlerDelegate = new MIL_GRA_HOOK_FUNCTION_PTR(HookHandler);
            MIL.MgraHookFunction(MilGraphicsList, MIL.M_GRAPHIC_MODIFIED, HookHandlerDelegate, GCHandle.ToIntPtr(DataStructureHandle));

            Console.WriteLine("You can try using your mouse or your keyboard to interactively");
            Console.WriteLine("modify the displayed region, such as moving, resizing, or");
            Console.WriteLine("rotating the region. If you do so, the results and annotations");
            Console.WriteLine("will be immediately updated.");
            Console.WriteLine();

            Console.WriteLine("Press <Enter> to exit.");

            int PressedKey = 0;
            while (PressedKey != ASCII_ENTER)
               PressedKey = (int)Console.ReadKey(true).Key;

            MIL.MgraHookFunction(MilGraphicsList, MIL.M_GRAPHIC_MODIFIED + MIL.M_UNHOOK, HookHandlerDelegate, GCHandle.ToIntPtr(DataStructureHandle));
            DataStructureHandle.Free();

            // Free all allocations.
            MIL.MblobFree(MilBlobResult);
            MIL.MblobFree(MilBlobContext);
            MIL.MbufFree(MilBinImage);
            MIL.MgraFree(MilGraphicsContext);
            MIL.MgraFree(MilGraphicsList);
            MIL.MbufFree(MilImage);
        }

        private static MIL_INT HookHandler(MIL_INT HookType, MIL_ID EventId, IntPtr UserDataPtr)
        {
            // this is how to check if the user data is null, the IntPtr class
            // contains a member, Zero, which exists solely for this purpose
            if (!IntPtr.Zero.Equals(UserDataPtr))
            {
                // get the handle to the DigHookUserData object back from the IntPtr
                GCHandle hUserData = GCHandle.FromIntPtr(UserDataPtr);

                // get a reference to the DigHookUserData object
                STestParameters DataStructure = hUserData.Target as STestParameters;

                // Check that the modified graphic is the rectangular region.
                MIL_INT ModifiedGraphicLabel = 0;
                MIL.MgraGetHookInfo(EventId, MIL.M_GRAPHIC_LABEL_VALUE, ref ModifiedGraphicLabel);

                if (ModifiedGraphicLabel == DataStructure.RegionLabel)
                {

                    // Count objects and draw the corresponding annotations.
                    CountObjects(DataStructure.MilDisplay,
                                 DataStructure.MilGraphicsList,
                                 DataStructure.MilGraphicsContext,
                                 DataStructure.MilBinImage,
                                 DataStructure.MilBlobContext,
                                 DataStructure.MilBlobResult);
                }
            }

            return MIL.M_NULL;
        }


        private const int MAX_TEXT_SIZE = 100;

        private static void CountObjects(MIL_ID MilDisplay, MIL_ID MilGraphicsList, MIL_ID MilGraphicsContext, MIL_ID MilBinImage, MIL_ID MilBlobContext, MIL_ID MilBlobResult)
        {
            MIL_INT NumberOfBlobs = 0;
            MIL_INT NumberOfPrimitives = 0;
            MIL_INT Index;

            string TextLabel;

            // Disable the display update for better performance.
            MIL.MdispControl(MilDisplay, MIL.M_UPDATE, MIL.M_DISABLE);

            // Remove all elements from the graphics list, except the rectangle
            // region primitive at index 0.
            MIL.MgraInquireList(MilGraphicsList, MIL.M_LIST, MIL.M_DEFAULT, MIL.M_NUMBER_OF_GRAPHICS, ref NumberOfPrimitives);
            for (Index = NumberOfPrimitives-1; Index > 0; Index--)
            {
                MIL.MgraControlList(MilGraphicsList, MIL.M_GRAPHIC_INDEX(Index), MIL.M_DEFAULT, MIL.M_DELETE, MIL.M_DEFAULT);
            }

            // Set the input region. The blob analysis will be done
            // from the (filled) interactive rectangle.
            MIL.MbufSetRegion(MilBinImage, MilGraphicsList, MIL.M_DEFAULT, MIL.M_RASTERIZE + MIL.M_FILL_REGION + MIL.M_USE_LINE_THICKNESS_1, MIL.M_DEFAULT);

            // Calculate the blobs and their features.
            MIL.MblobCalculate(MilBlobContext, MilBinImage, MIL.M_NULL, MilBlobResult);

            // Get the total number of blobs.
            MIL.MblobGetResult(MilBlobResult, MIL.M_GENERAL, MIL.M_NUMBER + MIL.M_TYPE_MIL_INT, ref NumberOfBlobs);

            // Set the input units to display unit for the count annotations.
            MIL.MgraControl(MilGraphicsContext, MIL.M_INPUT_UNITS, MIL.M_DISPLAY);
            TextLabel = string.Format(" Number of blobs found: {0,2} ", NumberOfBlobs);

            MIL.MgraColor(MilGraphicsContext, MIL.M_COLOR_WHITE);
            MIL.MgraText(MilGraphicsContext, MilGraphicsList, 10, 10, TextLabel);

            // Restore the input units to pixel units for result annotations.
            MIL.MgraControl(MilGraphicsContext, MIL.M_INPUT_UNITS, MIL.M_PIXEL);

            // Draw blob center of gravity annotation.
            MIL.MgraControl(MilGraphicsContext, MIL.M_LINE_THICKNESS, 3.0);
            MIL.MgraColor(MilGraphicsContext, MIL.M_COLOR_RED);
            MIL.MblobDraw(MilGraphicsContext, MilBlobResult, MilGraphicsList, MIL.M_DRAW_CENTER_OF_GRAVITY, MIL.M_INCLUDED_BLOBS, MIL.M_DEFAULT);

            // Draw blob bounding box annotations.
            MIL.MgraColor(MilGraphicsContext, MIL.M_COLOR_GREEN);
            MIL.MgraControl(MilGraphicsContext, MIL.M_LINE_THICKNESS, 1.0);
            MIL.MblobDraw(MilGraphicsContext, MilBlobResult, MilGraphicsList, MIL.M_DRAW_BOX, MIL.M_INCLUDED_BLOBS, MIL.M_DEFAULT);

            // Enable the display to update the drawings.
            MIL.MdispControl(MilDisplay, MIL.M_UPDATE, MIL.M_ENABLE);
        }
    }
}
